presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/ReservoirSampleFunction.java

```diff
@@ -0,0 +1,216 @@
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.metadata.BoundVariables;
+import com.facebook.presto.metadata.FunctionAndTypeManager;
+import com.facebook.presto.metadata.SqlAggregationFunction;
+import com.facebook.presto.operator.aggregation.AccumulatorCompiler;
+import com.facebook.presto.operator.aggregation.BuiltInAggregationFunctionImplementation;
+import com.facebook.presto.spi.function.AccumulatorState;
+import com.facebook.presto.spi.function.AccumulatorStateFactory;
+import com.facebook.presto.spi.function.AccumulatorStateSerializer;
+import com.facebook.presto.spi.function.aggregation.Accumulator;
+import com.facebook.presto.spi.function.aggregation.AggregationMetadata;
+import com.facebook.presto.spi.function.aggregation.GroupedAccumulator;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.invoke.MethodHandle;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.operator.aggregation.AggregationUtils.generateAggregationName;
+import static com.facebook.presto.spi.function.Signature.typeVariable;
+import static com.facebook.presto.spi.function.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.BLOCK_INDEX;
+import static com.facebook.presto.spi.function.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.INPUT_CHANNEL;
+import static com.facebook.presto.spi.function.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.NULLABLE_BLOCK_INPUT_CHANNEL;
+import static com.facebook.presto.spi.function.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.STATE;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.Math.max;
+
+public class ReservoirSampleFunction
+        extends SqlAggregationFunction
+{
+    public static final ReservoirSampleFunction RESERVOIR_SAMPLE = new ReservoirSampleFunction();
+    public static final String NAME = "reservoir_sample";
+    private static final MethodHandle INPUT_FUNCTION = methodHandle(ReservoirSampleFunction.class, "input", Type.class, ReservoirSampleState.class, Block.class, int.class, long.class, Block.class, int.class, long.class);
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(ReservoirSampleFunction.class, "combine", Type.class, ReservoirSampleState.class, ReservoirSampleState.class);
+    private static final MethodHandle OUTPUT_FUNCTION = methodHandle(ReservoirSampleFunction.class, "output", Type.class, ReservoirSampleState.class, BlockBuilder.class);
+
+    protected ReservoirSampleFunction()
```

Comment:
Can we use the annotations framework to define this function, [as specified in the documentation](https://prestodb.io/docs/current/develop/functions.html#aggregation-function-implementation)?

Reply:
Once #21316 is merged, I can update this implementation